src/conn/conn_capacity.c

```diff
@@ -0,0 +1,461 @@
+	 * recovering, we don't reach this code.
+	 */
+	if (type != WT_THROTTLE_READ) {
+		(void)__wt_atomic_addv64(&cap->written, bytes);
+		WT_STAT_CONN_INCRV(session, capacity_bytes_written, bytes);
+		__capacity_signal(session);
+	}
+
+	/* If we get sizes larger than this, later calculations may overflow. */
+	WT_ASSERT(session, bytes < 16 * (uint64_t)WT_GIGABYTE);
+	WT_ASSERT(session, capacity != 0);
+
+	/* Get the current time in nanoseconds since the epoch. */
+	__wt_epoch(session, &now);
+	now_ns = (uint64_t)now.tv_sec * WT_BILLION + (uint64_t)now.tv_nsec;
+
+again:
+	/* Take a reservation for the subsystem, and for the total */
+	__capacity_reserve(session,
+	    reservation, bytes, capacity, now_ns, &res_value);
+	__capacity_reserve(session, &cap->reservation_total, bytes,
+	    total_capacity, now_ns, &res_total_value);
+
+	/*
+	 * If we ended up with a future reservation, and we aren't constricted
+	 * by the total capacity, then we may be able to reallocate some
+	 * unused reservation time from another subsystem.
+	 */
+	if (res_value > now_ns && res_total_value < now_ns && steal == NULL &&
+	    total_capacity != 0) {
+		best_res = now_ns - WT_BILLION / 2;
+		if (type != WT_THROTTLE_CKPT &&
+		    (this_res = cap->reservation_ckpt) < best_res) {
+			steal = &cap->reservation_ckpt;
+			steal_capacity = cap->ckpt;
+			best_res = this_res;
+		}
+		if (type != WT_THROTTLE_EVICT &&
+		    (this_res = cap->reservation_evict) < best_res) {
+			steal = &cap->reservation_evict;
+			steal_capacity = cap->evict;
+			best_res = this_res;
+		}
+		if (type != WT_THROTTLE_LOG &&
+		    (this_res = cap->reservation_log) < best_res) {
+			steal = &cap->reservation_log;
+			steal_capacity = cap->log;
+			best_res = this_res;
+		}
+		if (type != WT_THROTTLE_READ &&
+		    (this_res = cap->reservation_read) < best_res) {
+			steal = &cap->reservation_read;
+			steal_capacity = cap->read;
+			best_res = this_res;
+		}
+
+		if (steal != NULL) {
+			/*
+			 * We have a subsystem that has enough spare capacity
+			 * to steal.  We'll take a small slice and add it to
+			 * our own subsystem.
+			 */
+			if (best_res < now_ns - WT_BILLION &&
+			    now_ns > WT_BILLION)
+				new_res = now_ns - WT_BILLION;
+			else
+				new_res = best_res;
+			WT_ASSERT(session, steal_capacity != 0);
+			new_res += WT_BILLION / 16 +
+			    WT_RESERVATION_NS(bytes, steal_capacity);
+			if (!__wt_atomic_casv64(steal, best_res, new_res)) {
+				/*
+				 * Give up our reservations and try again.
+				 * We won't try to steal the next time.
+				 */
+				(void)__wt_atomic_sub64(reservation,
+				    WT_RESERVATION_NS(bytes, capacity));
+				(void)__wt_atomic_sub64(&cap->reservation_total,
+				    WT_RESERVATION_NS(bytes, total_capacity));
+				goto again;
+			}
+
+			/*
+			 * We've actually stolen capacity in terms of bytes,
+			 * not nanoseconds, so we need to convert it.
+			 */
+			stolen_bytes = steal_capacity / 16;
```

Comment:
I'm a little confused why dividing by 16 converts bytes to nanoseconds?

Maybe a `#define` plus comment?

Reply:
@ddanderson please add this comment. I forget why this works, so I think a comment will be helpful to remember in the future.

Reply:
I've added clarifying comments and a macro.